pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/IndexCreator.java

```diff
@@ -0,0 +1,91 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.pinot.segment.spi.index;
+
+import java.io.Closeable;
+import java.io.IOException;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+
+/**
+ * The interface used to create indexes.
+ *
+ * The lifecycle for an IndexCreator is to be created, receive one or more calls to either
+ * {@link #addSingleValueCell(Object, int)} or {@link #addMultiValueCell(Object[], int[])} (but not mix them),
+ * a call to {@link #seal()} and finally be closed. Calls to add cell methods must be done in document id order,
+ * starting from the first document id.
+ */
+public interface IndexCreator extends Closeable {
+
+  /**
+   * This method can be implemented to offer a different value to the raw one included in the row source.
+   *
+   * Some specific indexes may need to change the actual value to be stored. For example, text indexes may offer a
+   * default value when {@link org.apache.pinot.spi.config.table.FieldConfig#TEXT_INDEX_NO_RAW_DATA} is defined.
+   *
+   * @param value the raw value included in the row
+   * @return Null in case this creator does not offer a new value.
+   */
+  @Nullable
+  default Object alternativeSingleValue(Object value) {
```

Comment:
This is anti-pattern, and we are trying to solve the current text index value alternate with disabled forward index. After that is done, we can remove the alternate APIs.
cc @siddharthteotia